File: typescript/src/roundtrip.ts
import { exchange, signHash, getMid } from "./client";

const COIN = "BTC";

async function sendOrder(side: string, px: string, sz: string) {
  const res = await exchange({
    action: {
      type: "order",
      orders: [{ asset: COIN, side, price: px, size: sz, tif: "ioc" }],
    },
  });
  const sig = await signHash(res.hash);

  return exchange({
    action: res.action,
    nonce: res.nonce,
    signature: sig,
  });
}

function checkStatuses(resp: any, label: string): boolean {
  const statuses = resp?.response?.data?.statuses || [];
  for (const s of statuses) {
    if (s.error) {
      console.error(`${label} error: ${s.error}`);
      return false;
    }
  }
  return true;
}

async function main() {
  const mid = await getMid(COIN);
  if (mid === 0) {
    console.error(`Could not fetch ${COIN} mid price`);
    process.exit(1);
  }

  const sz = (11.0 / mid).toFixed(5);
  console.log(`${COIN} mid: $${mid.toLocaleString()}`);
  console.log(`Trade size: ${sz} ${COIN} (~$${(parseFloat(sz) * mid).toFixed(2)})\n`);

  const buyPx = Math.floor(mid * 1.03).toString();
  console.log(`BUY ${sz} @ ${buyPx} (IOC)`);
  const buyResult = await sendOrder("buy", buyPx, sz);
  const buyResp = buyResult.exchangeResponse;
  if (!checkStatuses(buyResp, "BUY")) {
    process.exit(1);
  }
  console.log(`Buy filled: ${JSON.stringify(buyResp, null, 2)}\n`);

  await new Promise((r) => setTimeout(r, 1000));

  const sellPx = Math.floor(mid * 0.97).toString();
  console.log(`SELL ${sz} @ ${sellPx} (IOC)`);
  const sellResult = await sendOrder("sell", sellPx, sz);
  const sellResp = sellResult.exchangeResponse;
  if (!checkStatuses(sellResp, "SELL")) {
    process.exit(1);
  }
  console.log(`Sell filled: ${JSON.stringify(sellResp, null, 2)}`);
  console.log("\nRound-trip complete. Position should be flat.");
}

main();

File: python/roundtrip.py
"""IOC buy then sell on BTC perp — full trade cycle (~$11 notional)."""

import json
import time
from client import exchange, sign_hash, get_mid

COIN = "BTC"

mid = get_mid(COIN)
if mid == 0:
    print(f"Could not fetch {COIN} mid price")
    raise SystemExit(1)

sz = round(11.0 / mid, 5)
print(f"{COIN} mid: ${mid:,.2f}")
print(f"Trade size: {sz} {COIN} (~${sz * mid:.2f})\n")


def send_order(side, px):
    res = exchange({
        "action": {
            "type": "order",
            "orders": [{
                "asset": COIN,
                "side": side,
                "price": str(px),
                "size": str(sz),
                "tif": "ioc",
            }],
        },
    })
    sig = sign_hash(res["hash"])
    return exchange({
        "action": res["action"],
        "nonce": res["nonce"],
        "signature": sig,
    })


buy_px = int(mid * 1.03)
print(f"BUY {sz} @ {buy_px} (IOC)")
buy_result = send_order("buy", buy_px)
buy_resp = buy_result["exchangeResponse"]
statuses = buy_resp.get("response", {}).get("data", {}).get("statuses", [])
for s in statuses:
    if isinstance(s, dict) and "error" in s:
        print(f"Buy error: {s['error']}")
        raise SystemExit(1)

print(f"Buy filled: {json.dumps(buy_resp, indent=2)}\n")
time.sleep(1)

sell_px = int(mid * 0.97)
print(f"SELL {sz} @ {sell_px} (IOC)")
sell_result = send_order("sell", sell_px)
sell_resp = sell_result["exchangeResponse"]
statuses = sell_resp.get("response", {}).get("data", {}).get("statuses", [])
for s in statuses:
    if isinstance(s, dict) and "error" in s:
        print(f"Sell error: {s['error']}")
        raise SystemExit(1)

print(f"Sell filled: {json.dumps(sell_resp, indent=2)}")
print("\nRound-trip complete. Position should be flat.")
